core/chains/evm/testutils/client.go

```diff
@@ -146,97 +146,98 @@ func (ts *testWSServer) newWSHandler(chainID *big.Int, callback JSONRPCHandler)
-				}
-				responseBatch := fmt.Sprintf("[%s]", strings.Join(responses, ","))
-				ts.t.Logf("Sending batch response: %v", responseBatch)
-				ts.mu.Lock()
-				err = conn.WriteMessage(websocket.BinaryMessage, []byte(responseBatch))
-				ts.mu.Unlock()
-				if err != nil {
-					ts.t.Logf("Failed to write message: %v", err)
-				}
-				return
-			}
-			// Handle single request
-			if e := req.Get("error"); e.Exists() {
-				ts.t.Logf("Received jsonrpc error: %v", e)
-				continue
-			}
+		}
+	}
+}
 
-			m := req.Get("method")
-			if m.Type != gjson.String {
-				ts.t.Logf("Method must be string: %v", m.Type)
-				return
-			}
+func (ts *testWSServer) handleNewMsg(chainID *big.Int, conn *websocket.Conn, callback JSONRPCHandler) error {
+	_, data, err := conn.ReadMessage()
+	if err != nil {
+		return err
+	}
 
-			var resp JSONRPCResponse
-			if chainID != nil && m.String() == "eth_chainId" {
-				resp.Result = `"0x` + chainID.Text(16) + `"`
-			} else if m.String() == "eth_syncing" {
-				resp.Result = "false"
-			} else {
-				resp = callback(m.String(), req.Get("params"))
-			}
-			id := req.Get("id")
-			var msg string
-			if resp.Error.Message != "" {
-				msg = fmt.Sprintf(`{"jsonrpc":"2.0","id":%s,"error":{"code":%d,"message":"%s"}}`, id, resp.Error.Code, resp.Error.Message)
-			} else {
-				msg = fmt.Sprintf(`{"jsonrpc":"2.0","id":%s,"result":%s}`, id, resp.Result)
-			}
-			ts.t.Logf("Sending message: %v", msg)
-			ts.mu.Lock()
-			err = conn.WriteMessage(websocket.BinaryMessage, []byte(msg))
-			ts.mu.Unlock()
+	ts.t.Log("Received message", string(data))
+
+	req := gjson.ParseBytes(data)
+
+	if req.IsArray() { // Handle batch request
+		ts.t.Log("Received batch request")
+		var responses []string
+		for i, reqElem := range req.Array() {
+			var response string
+			response, _, err = ts.handleRequest(chainID, callback, reqElem)
 			if err != nil {
-				ts.t.Logf("Failed to write message: %v", err)
-				return
+				return fmt.Errorf("failed to handle elem %d of batch request: %w", i, err)
 			}
+			responses = append(responses, response)
+		}
 
-			if resp.Notify != "" {
-				time.Sleep(100 * time.Millisecond)
-				msg := fmt.Sprintf(`{"jsonrpc":"2.0","method":"eth_subscription","params":{"subscription":"0x00","result":%s}}`, resp.Notify)
-				ts.t.Log("Sending message", msg)
-				ts.mu.Lock()
-				err = conn.WriteMessage(websocket.BinaryMessage, []byte(msg))
-				ts.mu.Unlock()
-				if err != nil {
-					ts.t.Logf("Failed to write message: %v", err)
-					return
-				}
-			}
+		return ts.writeMsg(conn, fmt.Sprintf("[%s]", strings.Join(responses, ",")))
+	}
+	// Handle single request
+	response, asyncResponse, err := ts.handleRequest(chainID, callback, req)
+	if err != nil {
+		return fmt.Errorf("failed to handle request: %w", err)
+	}
+
+	if response != "" {
+		err = ts.writeMsg(conn, response)
+		if err != nil {
+			return err
 		}
 	}
+
+	if asyncResponse != "" {
+		time.Sleep(100 * time.Millisecond)
```

Comment:
Why do we need `time.Sleep` here?

Reply:
That's a good question. It was in the original implementation, and I was not sure if it was a good idea to remove it.
My guess is that we add delay in an attempt to simulate the async nature of subscription notifications.